### PR TITLE
[CALCITE-4532] Correct code generated for primitive-object ConstantExpression

### PR DESCRIPTION
**Pull request by gierlachg**
[CALCITE-4532](https://issues.apache.org/jira/browse/CALCITE-4532)

### CODE REVIEW

File: linq4j/src/main/java/org/apache/calcite/linq4j/tree/ConstantExpression.java [hunk 1]
```diff
@@ -136,6 +136,10 @@ private static ExpressionWriter write(ExpressionWriter writer,
       write(writer, value, primitive2.primitiveClass);
       return writer.append(")");
     }
+    Primitive primitive3 = Primitive.ofBox(value.getClass());
+    if (Object.class.equals(type) && primitive3 != null) {
```
Comment (zinking) [inline linq4j/src/main/java/org/apache/calcite/linq4j/tree/ConstantExpression.java:140]: doesn't primitive just mean unboxed java type ? can you call it just boxed ?

Comment (gierlachg) [reply linq4j/src/main/java/org/apache/calcite/linq4j/tree/ConstantExpression.java:140]: Wanted to stay consistent with `primitive2` couple of lines above.

**Comment by rubenada** [comment]
@DonnyZone any chance you could take a quick look at this one?

**Comment by DonnyZone** [comment]
> @DonnyZone any chance you could take a quick look at this one?

Sure, I will take a look today.

File: linq4j/src/test/java/org/apache/calcite/linq4j/test/ExpressionTest.java [hunk 2]
```diff
@@ -1357,6 +1357,25 @@ public void checkBlockBuilder(boolean optimizing, String expected) {
...
+    assertEquals("100",
+        Expressions.toString(Expressions.constant(100, Object.class)));
+    assertEquals("100L",
+        Expressions.toString(Expressions.constant(100L, Object.class)));
+    assertEquals("100.0F",
+        Expressions.toString(Expressions.constant(100F, Object.class)));
+    assertEquals("100.0D",
+        Expressions.toString(Expressions.constant(100D, Object.class)));
+  }
+
```
Comment (DonnyZone) [inline linq4j/src/test/java/org/apache/calcite/linq4j/test/ExpressionTest.java:1378]: The following two test case also pass without this commit. We can remove them.
```
    assertEquals("true",
        Expressions.toString(Expressions.constant(true, Object.class)));
    assertEquals("100",
        Expressions.toString(Expressions.constant(100, Object.class)));
```

Comment (gierlachg) [reply linq4j/src/test/java/org/apache/calcite/linq4j/test/ExpressionTest.java:1374]: Done.

**Comment by DonnyZone** [comment]
LGTM

**Review by DonnyZone — APPROVED** [review]

**Comment by DonnyZone** [comment]
Hi @gierlachg, could you please squash the commits? So that I can merge your fix through Github UI quickly.
My local environment has encountered some token problems today.

**Comment by gierlachg** [comment]
Of course. Just did it.

**Pull request merged**